### PR TITLE
feat: decrease the `page size` if the response size exceeds the limit

### LINKED ISSUE #2680: Reduce the `PAGE_SIZE` if the response size reaches limitation of grpc message
### What type of enhancement is this?

Refactor

### What does the enhancement do?

The range query's response message size might exceed the limit of grpc message; `PaginationStream` should able to reduce the `page_size` automatically.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Decreases the `page size` if the response size exceeds the limit

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#close #2680

### CODE REVIEW

**Review by MichaelScofield — APPROVED** [review]

File: src/common/meta/src/range_stream.rs [hunk 1]
```diff
@@ -92,11 +92,29 @@ impl PaginationStreamFactory {
...
-    pub async fn read_next(self) -> Result<(Self, Option<RangeResponse>)> {
+    /// Decrease the `page size` if the response message size exceeds the limitation.
+    /// TODO(weny): Considers to add an E2e test.
+    #[async_recursion::async_recursion]
+    async fn adaptive_range(&mut self, req: RangeRequest) -> Result<RangeResponse> {
+        match self.kv.range(req.clone()).await {
+            Ok(resp) => Ok(resp),
+            Err(err) => {
+                if err.is_exceeded_size_limit() {
+                    self.page_size /= 2;
```
Comment (fengjiachun) [inline src/common/meta/src/range_stream.rs:103]: We need to make sure that page_size >= 1

Comment (fengjiachun) [inline src/common/meta/src/range_stream.rs:103]: We need to make sure that page_size >= 1

File: src/common/meta/src/range_stream.rs [hunk 2]
```diff
@@ -87,16 +91,59 @@ impl PaginationStreamFactory {
...
+                error::UnexpectedSnafu {
+                    err_msg: "Exceeded maximum number of adaptive range retries"
+                }
+            );
+            self.adaptive_page_size = Some(new_adaptive_page_size);
+        } else {
+            self.adaptive_page_size = Some(if self.page_size == 0 {
+                DEFAULT_ADAPTIVE_PAGE_SIZE
+            } else {
+                self.page_size / 2
```
Comment (fengjiachun) [inline src/common/meta/src/range_stream.rs:116]: if `page_size` == 1, than `adaptive_page_size` = 0

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**